azure/services/publicips/publicips.go

```diff
@@ -57,91 +59,75 @@ func (s *Service) Reconcile(ctx context.Context) error {
+	var result error
 
-		// only set DNS properties if there is a DNS name specified
-		var dnsSettings *network.PublicIPAddressDNSSettings
-		if ip.DNSName != "" {
-			dnsSettings = &network.PublicIPAddressDNSSettings{
-				DomainNameLabel: to.StringPtr(strings.Split(ip.DNSName, ".")[0]),
-				Fqdn:            to.StringPtr(ip.DNSName),
+	for _, ipSpec := range s.Scope.PublicIPSpecs() {
+		if err := async.CreateResource(ctx, s.Scope, s.Client, ipSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
 			}
 		}
-
-		err := s.Client.CreateOrUpdate(
-			ctx,
-			s.Scope.ResourceGroup(),
-			ip.Name,
-			network.PublicIPAddress{
-				Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-					ClusterName: s.Scope.ClusterName(),
-					Lifecycle:   infrav1.ResourceLifecycleOwned,
-					Name:        to.StringPtr(ip.Name),
-					Additional:  s.Scope.AdditionalTags(),
-				})),
-				Sku:      &network.PublicIPAddressSku{Name: network.PublicIPAddressSkuNameStandard},
-				Name:     to.StringPtr(ip.Name),
-				Location: to.StringPtr(s.Scope.Location()),
-				PublicIPAddressPropertiesFormat: &network.PublicIPAddressPropertiesFormat{
-					PublicIPAddressVersion:   addressVersion,
-					PublicIPAllocationMethod: network.IPAllocationMethodStatic,
-					DNSSettings:              dnsSettings,
-				},
-				Zones: to.StringSlicePtr(s.Scope.FailureDomains()),
-			},
-		)
-
-		if err != nil {
-			return errors.Wrap(err, "cannot create public IP")
-		}
-
-		s.Scope.V(2).Info("successfully created public IP", "public ip", ip.Name)
 	}
 
-	return nil
+	s.Scope.UpdatePutStatus(infrav1.PublicIPsReadyCondition, serviceName, result)
+	return result
 }
 
 // Delete deletes the public IP with the provided scope.
 func (s *Service) Delete(ctx context.Context) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "publicips.Service.Delete")
 	defer done()
 
-	for _, ip := range s.Scope.PublicIPSpecs() {
-		managed, err := s.isIPManaged(ctx, ip.Name)
-		if err != nil && !azure.ResourceNotFound(err) {
-			return errors.Wrap(err, "could not get public IP management state")
-		}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-		if !managed {
-			s.Scope.V(2).Info("Skipping IP deletion for unmanaged public IP", "public ip", ip.Name)
+	// We go through the list of public ip specs to delete each one, independently of the result of the previous one.
+	// If multiple errors occur, we return the most pressing one
+	// order of precedence is: error deleting -> error getting management state -> deleting in progress -> deleted (no error)
+	// TODO(karuppiah7890): Is the above precedence okay? Ask maintainers. Accordingly make changes in test and code.
+	// Ask maintainers what's the precedence for - error getting public IP management state. Is it equal to the error deleting IP or less than or more than that?
```

Comment:
Great question. I would say the error importance for error getting management state is equal to a deleting IP error since in the end they are both errors trying to delete the IP (the only one that "less urgent" is the operation not done error, which just means the delete is in progress, so we want to prioritize actual errors before showing not done errors in Status conditions). For errors getting management state, you can treat them the same as delete errors, ie. the last one is the one returned.

Reply:
In the end we're still going to keep retrying all the operations until there are no errors left, but the error precedence is used to show the user the most pressing one in Status.

Reply:
Cool ! I'll make changes accordingly ! Also, since there are multiple public IPs and we show just one status for the whole set of public IPs, do we want to show the error status of just one of the public IPs when there are multiple errors from multiple public IPs?

Reply:
Yeah that's what I've been doing for the other services so far but we haven't merged any with multiple resources yet (only open PRs #1684, #1686 for now). My reasoning is is that this is used as "reason for failing and marking as   not ready" so whether there is 1 or more it doesn't matter, in the end the first one (or last one) is ultimately what causes it to fail and retry. If you have any suggestions for showing more than 1 in the Status in an elegant way let me know! Also cc @Jont828 whom I've been discussing this with.

Reply:
> If you have any suggestions for showing more than 1 in the Status in an elegant way let me know!

I gotta think on that. But yeah, many errors will be long - with more details. I was thinking something basic - if error is a one liner, show multiple lines separated by dot `.` and add public IP name in front of the error to show which public IPs are having errors. I'll come up with an example sometime and discuss about it. Maybe that could be a separate discussion and PR by itself and something to consider for other resources too, where multiple resources are handled in a single reconcile

Reply:
> and something to consider for other resources too, where multiple resources are handled in a single reconcile and reported with a single status

definitely, whatever we end up doing should be consistent across resource types

Reply:
Implemented what's been mentioned here for same error precedence for error getting management state and creation error. Created issue #1900 to discuss showing multiple errors